src/session/session_helper.c

```diff
@@ -159,24 +160,44 @@ __wt_session_set_last_error(
     if (!F_ISSET(session, WT_SESSION_SAVE_ERRORS))
         return (0);
 
-    /* Format the error message string. */
-    WT_RET(__wt_scr_alloc(session, 0, &buf));
-    WT_VA_ARGS_BUF_FORMAT(session, buf, fmt, false);
-    err_msg_content = buf->data;
+    /*
+     * If the error code is 0, we don't bother allocating memory for the message, as we have static
+     * buffers for the two possible success messages (WT_ERROR_INFO_EMPTY/WT_ERROR_INFO_SUCCESS).
+     *
+     * Otherwise, if the error message is the empty string, we also use the static buffer
+     * WT_ERROR_INFO_EMPTY rather than allocating.
+     */
+    if (!((new_msg_is_empty = strcmp(fmt, WT_ERROR_INFO_EMPTY) == 0) || err == 0)) {
+
+        /* Format the error message string. */
+        WT_RET(__wt_scr_alloc(session, 0, &buf));
+        WT_VA_ARGS_BUF_FORMAT(session, buf, fmt, false);
+        err_msg_content = buf->data;
 
-    /* Only set the error if it results in a change. */
-    if (err_info->err == err && err_info->sub_level_err == sub_level_err &&
-      err_info->err_msg != NULL && strcmp(err_info->err_msg, err_msg_content) == 0)
-        goto err;
+        /* Return early if the new error is identical to the previously stored error. */
+        if (err_info->err == err && err_info->sub_level_err == sub_level_err &&
+          err_info->err_msg != NULL && strcmp(err_info->err_msg, err_msg_content) == 0)
+            goto err;
+    }
 
-    /* Free the last error message string. */
-    __wt_free(session, err_info->err_msg);
+    /* Free the last saved error message string if it was dynamically allocated. */
+    if (err_info->err != 0 && err_info->err_msg != NULL &&
+      strcmp(err_info->err_msg, WT_ERROR_INFO_EMPTY) != 0)
+        __wt_free(session, err_info->err_msg);
 
-    /* Load error codes and message content into err_info. */
-    WT_ERR(__wt_calloc(session, buf->size + 1, 1, &(err_info->err_msg)));
-    WT_ERR(__wt_snprintf(err_info->err_msg, buf->size + 1, "%s", err_msg_content));
+    /* Load error codes and message into err_info. */
     err_info->err = err;
     err_info->sub_level_err = sub_level_err;
+    if (new_msg_is_empty) {
+        err_info->err_msg = WT_ERROR_INFO_EMPTY;
+    } else if (err == 0) {
+        err_info->err_msg = WT_ERROR_INFO_SUCCESS;
+    } else {
```

Comment:
Style: no need for curly brackets for one liners
```suggestion
    if (new_msg_is_empty)
        err_info->err_msg = WT_ERROR_INFO_EMPTY;
    else if (err == 0)
        err_info->err_msg = WT_ERROR_INFO_SUCCESS;
    else {
```